Sources/Haversack/Haversack.swift

```diff
@@ -344,170 +470,67 @@ public struct Haversack {
 
     private func unsynchronizedKeyGeneration(fromConfig config: KeyGenerationConfig,
                                              itemSecurity: ItemSecurity) throws -> SecKey {
-        var query = self.merge(keyConfig: config, withSecurity: itemSecurity)
-
-        #if os(macOS)
-            query = try self.addKeychain(to: query, forAdd: true)
-        #endif
-
-        try self.precheck(query)
+        let query = try makeKeyGenerationQuery(fromConfig: config, itemSecurity: itemSecurity)
 
         return try self.configuration.strategy.generateKey(query)
     }
 
     private func unsynchronizedDelete<T: KeychainStorable>(_ item: T, treatNotFoundAsSuccess: Bool) throws {
-        var localQuery = item.entityQuery(includeSecureData: false)
-
-        #if os(macOS)
-            localQuery = try self.addKeychain(to: localQuery, forAdd: false)
-            // iOS does not support kSecMatchLimit for delete operations
-            localQuery[kSecMatchLimit as String] = kSecMatchLimitOne
-        #endif
+        let localQuery = try makeDeleteQuery(item)
 
         try self.configuration.strategy.delete(localQuery, treatNotFoundAsSuccess: treatNotFoundAsSuccess)
     }
+}
 
-    // MARK: - Deletion
-
-    /// Synchronously delete an item from the keychain that was previously retrieved from the keychain.
-    ///
-    /// If the item does not include a `reference` previously retrieved from the keychain: on iOS/tvOS/visionOS/watchOS
-	/// all items matching the item metadata will be deleted, while on macOS only the first matching item will be deleted.
-    /// - Parameter item: The item retrieved from the keychain.
-    /// - Parameter treatNotFoundAsSuccess: If true, no error is thrown when the query does not
-    /// find an item to delete; default is true.
-    /// - Throws: A ``HaversackError`` object if any errors occur.
-    public func delete<T: KeychainStorable>(_ item: T, treatNotFoundAsSuccess: Bool = true) throws {
-        return try configuration.serialQueue.sync {
-            try self.unsynchronizedDelete(item, treatNotFoundAsSuccess: treatNotFoundAsSuccess)
-        }
+// MARK: Query builders
+extension Haversack {
+    package func makeSearchQuery<T: KeychainQuerying>(_ query: T, singleItem: Bool) throws -> T {
+        try precheckSearch(query, singleItem: singleItem)
```

Comment:
Please move the contents of `precheckSearch` into this function and delete `precheckSearch`.  This is the only place that `precheckSearch` is called.